### PR TITLE
NDRS-484: Replace Serde with ToBytes/FromBytes implementation

### PR DESCRIPTION
**Pull request by darthsiroftardis**
Replaced usage of `serde` with our custom serialization implementation for data structures interacting with the storage component. 

REF: https://casperlabs.atlassian.net/browse/NDRS-484

### CODE REVIEW

File: node/src/components/storage.rs [hunk 1]
```diff
@@ -113,35 +119,64 @@ pub trait WithBlockHeight: Value {
...
         let _ = execution_results.insert(block_hash, execution_result);
         DeployMetadata { execution_results }
     }
 }
 
+#[cfg(test)]
+impl DeployMetadata<Block> {
+    fn random(rng: &mut TestRng) -> Self {
+        let block = Block::random(rng);
+        let id = super::storage::Value::id(&block);
```
Comment (Fraser999) [inline node/src/components/storage.rs:148]:
```suggestion
        let id = Value::id(&block);
```

File: node/src/components/storage.rs [hunk 2]
```diff
@@ -113,35 +119,64 @@ pub trait WithBlockHeight: Value {
...
+
 impl<B: Value> Default for DeployMetadata<B> {
     fn default() -> Self {
         DeployMetadata {
-            execution_results: HashMap::new(),
+            execution_results: BTreeMap::new(),
         }
     }
 }
+impl<B: Value> ToBytes for DeployMetadata<B> {
```
Comment (Fraser999) [inline node/src/components/storage.rs:163]:
```suggestion
}

impl<B: Value> ToBytes for DeployMetadata<B> {
```

File: node/src/components/storage.rs [hunk 3]
```diff
@@ -113,35 +119,64 @@ pub trait WithBlockHeight: Value {
...
+    fn to_bytes(&self) -> StdResult<Vec<u8>, bytesrepr::Error> {
+        let mut buffer = bytesrepr::allocate_buffer(self)?;
+        buffer.extend(self.execution_results.to_bytes()?);
+        Ok(buffer)
+    }
+    fn serialized_length(&self) -> usize {
```
Comment (Fraser999) [inline node/src/components/storage.rs:169]:
```suggestion
    }

    fn serialized_length(&self) -> usize {
```

File: node/src/components/storage.rs [hunk 4]
```diff
@@ -113,35 +119,64 @@ pub trait WithBlockHeight: Value {
...
+        self.execution_results.serialized_length()
+    }
+}
+impl<B: Value> FromBytes for DeployMetadata<B> {
```
Comment (Fraser999) [inline node/src/components/storage.rs:173]:
```suggestion
}

impl<B: Value> FromBytes for DeployMetadata<B> {
```

File: node/src/components/storage/lmdb_block_height_store.rs [hunk 5]
```diff
@@ -4,10 +4,11 @@ use std::{
     sync::atomic::{AtomicU64, Ordering},
 };
 
+use casper_types::bytesrepr::{self, FromBytes, ToBytes};
```
Comment (Fraser999) [inline node/src/components/storage/lmdb_block_height_store.rs:7]: The imports should be grouped as follows, with a blank line between each group:
* std
* third party libs
* CasperLabs libs other than the current crate
* the current crate

File: node/src/types/block.rs [hunk 6]
```diff
@@ -11,12 +11,14 @@ use blake2::{
...
     VarBlake2b,
 };
+use casper_types::bytesrepr::{self, FromBytes, ToBytes};
 use datasize::DataSize;
 use hex::FromHexError;
 use hex_fmt::{HexFmt, HexList};
 #[cfg(test)]
 use rand::Rng;
 use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
```
Comment (Fraser999) [inline node/src/types/block.rs:21]: These imports need sorted as per previous comment too.

File: node/src/types/block.rs [hunk 7]
```diff
@@ -705,6 +818,41 @@ impl Display for Block {
         Ok(())
     }
 }
+///        parent_hash: BlockHash,
+/// parent_seed: Digest,
+///  state_root_hash: Digest,
+/// finalized_block: FinalizedBlock,
```
Comment (Fraser999) [inline node/src/types/block.rs:824]: Left in by mistake?

File: node/src/types/deploy.rs [hunk 8]
```diff
@@ -295,9 +317,28 @@ pub struct Deploy {
...
-    /// Constructs a new `Deploy`.
+    /// Constructs a new unsigned `Deploy`
     #[allow(clippy::too_many_arguments)]
     pub fn new(
+        hash: DeployHash,
+        header: DeployHeader,
+        payment: ExecutableDeployItem,
+        session: ExecutableDeployItem,
+        approvals: Vec<Approval>,
+    ) -> Self {
```
Comment (Fraser999) [inline node/src/types/deploy.rs:328]: Let's revert this change since it seems like this `new()` constructor isn't used anywhere.

File: node/src/types/json_compatibility/execution_result.rs [hunk 9]
```diff
@@ -18,11 +19,37 @@ use casper_execution_engine::{
...
+};
 
 #[cfg(test)]
 use crate::testing::TestRng;
 
+/// Constants to track operation serialization.
+pub const OP_READ_TAG: u8 = 1;
+pub const OP_WRITE_TAG: u8 = 2;
+pub const OP_ADD_TAG: u8 = 3;
+pub const OP_NOOP_TAG: u8 = 4;
```
Comment (Fraser999) [inline node/src/types/json_compatibility/execution_result.rs:34]: None of these should be `pub`.  Also, let's zero-index them as per other cases.

File: node/src/types/json_compatibility/execution_result.rs [hunk 10]
```diff
@@ -18,11 +19,37 @@ use casper_execution_engine::{
...
+pub const WRITE_CONTRACT_PACKAGE: u8 = 5;
+pub const WRITE_DEPLOY_INFO: u8 = 6;
+pub const WRITE_TRANSFER: u8 = 7;
+pub const ADD_INT_I32: u8 = 8;
+pub const ADD_INT_U64: u8 = 9;
+pub const ADD_INT_U128: u8 = 10;
+pub const ADD_INT_U256: u8 = 11;
+pub const ADD_UINT_512: u8 = 12;
+pub const ADD_KEYS: u8 = 13;
+pub const FAILURE: u8 = 14;
```
Comment (Fraser999) [inline node/src/types/json_compatibility/execution_result.rs:51]: None of these should be `pub`.  Let's append `_TAG` to all of them as well.

File: node/src/types/json_compatibility/execution_result.rs [hunk 11]
```diff
@@ -18,11 +19,37 @@ use casper_execution_engine::{
...
+pub const WRITE_CONTRACT_WASM: u8 = 3;
+pub const WRITE_CONTRACT: u8 = 4;
+pub const WRITE_CONTRACT_PACKAGE: u8 = 5;
+pub const WRITE_DEPLOY_INFO: u8 = 6;
+pub const WRITE_TRANSFER: u8 = 7;
+pub const ADD_INT_I32: u8 = 8;
+pub const ADD_INT_U64: u8 = 9;
+pub const ADD_INT_U128: u8 = 10;
+pub const ADD_INT_U256: u8 = 11;
+pub const ADD_UINT_512: u8 = 12;
```
Comment (Fraser999) [inline node/src/types/json_compatibility/execution_result.rs:49]:
```suggestion
pub const ADD_INT32_TAG: u8 = 8;
pub const ADD_UINT64_TAG: u8 = 9;
pub const ADD_UINT128_TAG: u8 = 10;
pub const ADD_UINT256_TAG: u8 = 11;
pub const ADD_UINT512_TAG: u8 = 12;
```

File: node/src/types/json_compatibility/execution_result.rs [hunk 12]
```diff
@@ -137,6 +219,47 @@ impl From<&Op> for Operation {
...
+                buffer.insert(0, OP_WRITE_TAG);
+            }
+            Operation::Add => {
+                buffer.insert(0, OP_ADD_TAG);
+            }
+            Operation::NoOp => {
+                buffer.insert(0, OP_NOOP_TAG);
+            }
+        }
+        Ok(buffer)
```
Comment (Fraser999) [inline node/src/types/json_compatibility/execution_result.rs:239]:
```suggestion
        match self {
            Operation::Read => OP_READ_TAG.to_bytes(),
            Operation::Write => OP_WRITE_TAG.to_bytes(),
            Operation::Add => OP_ADD_TAG.to_bytes(),
            Operation::NoOp => OP_NOOP_TAG.to_bytes(),
        }
```

File: node/src/types/json_compatibility/execution_result.rs [hunk 13]
```diff
@@ -137,6 +219,47 @@ impl From<&Op> for Operation {
...
+    }
+
+    fn serialized_length(&self) -> usize {
+        bytesrepr::U8_SERIALIZED_LENGTH
```
Comment (Fraser999) [inline node/src/types/json_compatibility/execution_result.rs:243]: `U8_SERIALIZED_LENGTH` should be imported.

File: node/src/types/json_compatibility/execution_result.rs [hunk 14]
```diff
@@ -213,3 +336,165 @@ impl From<&EngineTransform> for Transform {
...
+            ADD_KEYS => {
+                let (value, remainder) = BTreeMap::<String, String>::from_bytes(remainder)?;
+                Ok((Transform::AddKeys(value), remainder))
+            }
+            FAILURE => {
+                let (value, remainder) = String::from_bytes(remainder)?;
+                Ok((Transform::Failure(value), remainder))
+            }
+            _ => {
+                info!("Failed to deserialize Transforms, invalid identifier found");
```
Comment (Fraser999) [inline node/src/types/json_compatibility/execution_result.rs:462]:
```suggestion
                info!("Failed to deserialize Transform, invalid identifier found");
```

**Review by Fraser999 — COMMENTED** [review]
We're still missing a serialization roundtrip test for `EraEnd` in block.rs.

Other than that and a few final nitpicks, it's looking good.

File: node/src/components/storage.rs (hunk 1 above)
Comment (darthsiroftardis) [reply node/src/components/storage.rs:148]: Value was being incorrectly flagged as unresolved. Changes made.

File: node/src/components/storage/lmdb_block_height_store.rs (hunk 5 above)
Comment (darthsiroftardis) [reply node/src/components/storage/lmdb_block_height_store.rs:7]: I blame VSCode for not knowing this convention. I have resolved it as requested.

File: node/src/types/json_compatibility/execution_result.rs (hunk 10 above)
Comment (darthsiroftardis) [reply node/src/types/json_compatibility/execution_result.rs]: Have done so

File: node/src/types/json_compatibility/execution_result.rs (hunk 11 above)
Comment (darthsiroftardis) [reply node/src/types/json_compatibility/execution_result.rs]: Renamed based on suggested change

File: node/src/types/json_compatibility/execution_result.rs (hunk 12 above)
Comment (darthsiroftardis) [reply node/src/types/json_compatibility/execution_result.rs]: Made requested change

File: node/src/types/json_compatibility/execution_result.rs (hunk 14 above)
Comment (darthsiroftardis) [reply node/src/types/json_compatibility/execution_result.rs]: Applied change

File: node/src/types/json_compatibility/execution_result.rs (hunk 13 above)
Comment (darthsiroftardis) [reply node/src/types/json_compatibility/execution_result.rs]: Imported accordingly

File: node/src/types/deploy.rs (hunk 8 above)
Comment (darthsiroftardis) [reply node/src/types/deploy.rs:328]: Will do

File: node/src/types/block.rs (hunk 7 above)
Comment (darthsiroftardis) [reply node/src/types/block.rs:824]: Yes :'(

File: node/src/types/block.rs (hunk 6 above)
Comment (darthsiroftardis) [reply node/src/types/block.rs:21]: Yessir

Comment (Fraser999) [reply node/src/types/block.rs]: Still got a stray `std` import

File: node/src/types/block.rs [hunk 15]
```diff
@@ -825,6 +973,22 @@ mod tests {
...
+        let mut rng = TestRng::new();
+        let block = Block::random(&mut rng);
+        bytesrepr::test_serialization_roundtrip(&block);
+    }
+
+    #[test]
+    fn bytesrepr_roundtrip_era_end() {
+        let mut rng = TestRng::new();
+        let finalized_block = FinalizedBlock::random(&mut rng);
+        if let Some(era_end) = finalized_block.era_end() {
```
Comment (Fraser999) [inline node/src/types/block.rs:987]: Let's loop, creating random finalized blocks until we get `Some(era_end)`.  That way we're at least guaranteed on every test run to be actually testing this roundtrip.

Comment (darthsiroftardis) [reply node/src/types/block.rs:987]: Aaah, I see your point

File: node/src/types/json_compatibility/execution_result.rs (hunk 9 above)
Comment (Fraser999) [reply node/src/types/json_compatibility/execution_result.rs]: This wasn't actioned, but you marked the issue resolved!

Comment (darthsiroftardis) [reply node/src/types/json_compatibility/execution_result.rs]: WELP

**Review by Fraser999 — APPROVED** [review]
